File: android/src/main/java/com/fidelreactlibrary/adapters/FidelSetupAdapter.java
package com.fidelreactlibrary.adapters;

import android.graphics.Bitmap;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.ReadableMap;
import com.fidelapi.Fidel;
import com.fidelapi.entities.Country;
import com.fidelreactlibrary.adapters.abstraction.CardSchemesAdapter;
import com.fidelreactlibrary.adapters.abstraction.ConstantsProvider;
import com.fidelreactlibrary.adapters.abstraction.CountryAdapter;
import com.fidelreactlibrary.adapters.abstraction.DataOutput;
import com.fidelreactlibrary.adapters.abstraction.DataProcessor;
import com.fidelreactlibrary.adapters.abstraction.ProgramTypeAdapter;

import org.json.JSONObject;

import java.util.Map;

public final class FidelSetupAdapter implements DataProcessor<ReadableMap>, DataOutput<Bitmap>, ConstantsProvider {

    private final DataProcessor<ReadableMap> imageAdapter;
    private final CountryAdapter countryAdapter;
    private final CardSchemesAdapter cardSchemesAdapter;
    private final ProgramTypeAdapter programTypeAdapter;

    public FidelSetupAdapter(DataProcessor<ReadableMap> imageAdapter, CountryAdapter countryAdapter,
            CardSchemesAdapter cardSchemesAdapter, ProgramTypeAdapter programTypeAdapter) {
        this.imageAdapter = imageAdapter;
        this.countryAdapter = countryAdapter;
        this.cardSchemesAdapter = cardSchemesAdapter;
        this.programTypeAdapter = programTypeAdapter;
    }

    @Override
    public void process(ReadableMap data) {
        if (data.hasKey(FidelSetupProperties.SDK_KEY.jsName())) {
            Fidel.sdkKey = data.getString(FidelSetupProperties.SDK_KEY.jsName());
        }
        if (data.hasKey(FidelSetupProperties.PROGRAM_ID.jsName())) {
            Fidel.programId = data.getString(FidelSetupProperties.PROGRAM_ID.jsName());
        }
        if (data.hasKey(FidelSetupProperties.PROGRAM_TYPE.jsName())) {
            String programTypeValue = data.getString(FidelSetupProperties.PROGRAM_TYPE.jsName());
            Fidel.programType = programTypeAdapter.parseProgramType(programTypeValue);
        }
        ReadableMap optionsMap = data.getMap(FidelSetupProperties.OPTIONS.jsName());
        if (optionsMap != null) {
            if (optionsMap.hasKey(FidelSetupProperties.Options.BANNER_IMAGE.jsName())) {
                imageAdapter.process(optionsMap.getMap(FidelSetupProperties.Options.BANNER_IMAGE.jsName()));
            }
            if (optionsMap.hasKey(FidelSetupProperties.Options.ALLOWED_COUNTRIES.jsName())) {
                Fidel.allowedCountries = countryAdapter
                        .parseAllowedCountries(optionsMap.getArray(FidelSetupProperties.Options.ALLOWED_COUNTRIES.jsName()));
            }
            if (optionsMap.hasKey(FidelSetupProperties.Options.DEFAULT_SELECTED_COUNTRY.jsName())) {
                Country defaultSelectedCountry = countryAdapter.countryWithJSValue(
                        optionsMap.getString(FidelSetupProperties.Options.DEFAULT_SELECTED_COUNTRY.jsName()));
                if (defaultSelectedCountry != null) {
                    Fidel.defaultSelectedCountry = defaultSelectedCountry;
                }
            }
            if (optionsMap.hasKey(FidelSetupProperties.Options.SUPPORTED_CARD_SCHEMES.jsName())) {
                Fidel.supportedCardSchemes = cardSchemesAdapter.cardSchemesWithReadableArray(
                        optionsMap.getArray(FidelSetupProperties.Options.SUPPORTED_CARD_SCHEMES.jsName()));
            }
            if (optionsMap.hasKey(FidelSetupProperties.Options.THIRD_PARTY_VERIFICATION_CHOICE.jsName())) {
                Fidel.thirdPartyVerificationChoice = optionsMap
                        .getBoolean(FidelSetupProperties.Options.THIRD_PARTY_VERIFICATION_CHOICE.jsName());
            }
            if (optionsMap.hasKey(FidelSetupProperties.Options.META_DATA.jsName())) {
                ReadableMap metaDataMap = optionsMap.getMap(FidelSetupProperties.Options.META_DATA.jsName());
                if (metaDataMap != null) {
                    Fidel.metaData = new JSONObject(metaDataMap.toHashMap());
                } else {
                    Fidel.metaData = null;
                }
            }
        }

        ReadableMap consentTextMap = data.getMap(FidelSetupProperties.CONSENT_TEXT.jsName());
        if (consentTextMap != null) {
            if (consentTextMap.hasKey(FidelSetupProperties.ConsentText.COMPANY_NAME.jsName())) {
                Fidel.companyName = consentTextMap.getString(FidelSetupProperties.ConsentText.COMPANY_NAME.jsName());
            }
            if (consentTextMap.hasKey(FidelSetupProperties.ConsentText.TERMS_AND_CONDITIONS_URL.jsName())) {
                Fidel.termsAndConditionsUrl = consentTextMap
                        .getString(FidelSetupProperties.ConsentText.TERMS_AND_CONDITIONS_URL.jsName());
            }
            if (consentTextMap.hasKey(FidelSetupProperties.ConsentText.PRIVACY_POLICY_URL.jsName())) {
                Fidel.privacyPolicyUrl = consentTextMap
                        .getString(FidelSetupProperties.ConsentText.PRIVACY_POLICY_URL.jsName());
            }
            if (consentTextMap.hasKey(FidelSetupProperties.ConsentText.PROGRAM_NAME.jsName())) {
                Fidel.programName = consentTextMap.getString(FidelSetupProperties.ConsentText.PROGRAM_NAME.jsName());
            }
            if (consentTextMap.hasKey(FidelSetupProperties.ConsentText.DELETE_INSTRUCTIONS.jsName())) {
                Fidel.deleteInstructions = consentTextMap
                        .getString(FidelSetupProperties.ConsentText.DELETE_INSTRUCTIONS.jsName());
            }
        }
    }

    @Override
    public void output(Bitmap data) {
        Fidel.bannerImage = data;
    }

    @NonNull
    @Override
    public Map<String, Object> getConstants() {
        Map<String, Object> constants = countryAdapter.getConstants();
        constants.putAll(cardSchemesAdapter.getConstants());
        constants.putAll(programTypeAdapter.getConstants());
        return constants;
    }
}

File: ios/Adapters/FidelSetupAdapter.swift
//
//  FidelSetupAdapter.swift
//  fidel-react-native
//
//  Created by Corneliu Chitanu on 08/12/21.
//

import Foundation
import Fidel

class FidelSetupAdapter: NSObject {

    private let imageAdapter: FLRNImageFromRNAdapter

    init(imageAdapter: FLRNImageFromRNAdapter) {
        self.imageAdapter = imageAdapter
        super.init()
    }

    func setup(with parameters: NSDictionary) {
        guard let parameters = parameters as? [String: Any?] else {
            return
        }
        if parameters.keys.contains(FidelSetupProperties.sdkKey.rawValue) {
            Fidel.sdkKey = parameters[FidelSetupProperties.sdkKey.rawValue] as? String
        }
        if parameters.keys.contains(FidelSetupProperties.programID.rawValue) {
            Fidel.programID = parameters[FidelSetupProperties.programID.rawValue] as? String
        }
        if parameters.keys.contains(FidelSetupProperties.programType.rawValue) {
            let programTypeJSValue = parameters[FidelSetupProperties.programType.rawValue] as? String
            Fidel.programType = ProgramType.programType(from: programTypeJSValue)
        }
        if parameters.keys.contains(FidelSetupProperties.options.rawValue),
           let options = parameters[FidelSetupProperties.options.rawValue] as? [String: Any?] {
            setOptions(options)
        }
        if parameters.keys.contains(FidelSetupProperties.consentText.rawValue),
           let consentTextValues = parameters[FidelSetupProperties.consentText.rawValue] as? [String: Any?] {
            setConsentTextValues(consentTextValues)
        }
    }

    func setOptions(_ options: [String: Any?]) {
        if options.keys.contains(FidelSetupProperties.Options.bannerImage.rawValue),
           let rawImageObject = options[FidelSetupProperties.Options.bannerImage.rawValue] as? NSObject {
            Fidel.bannerImage = imageAdapter.image(fromRawData: rawImageObject)
        }
        if options.keys.contains(FidelSetupProperties.Options.allowedCountries.rawValue),
           let countryJSKeys = options[FidelSetupProperties.Options.allowedCountries.rawValue] as? [String] {
            Fidel.allowedCountries = Country.countriesSet(from: countryJSKeys)
        }
        if options.keys.contains(FidelSetupProperties.Options.defaultSelectedCountry.rawValue),
           let countryJSKey = options[FidelSetupProperties.Options.defaultSelectedCountry.rawValue] as? String,
           let defaultSelectedCountry = Country.country(from: countryJSKey) {
            Fidel.defaultSelectedCountry = defaultSelectedCountry
        }
        if options.keys.contains(FidelSetupProperties.Options.supportedCardSchemes.rawValue),
           let supportedCardSchemeJSKeys = options[FidelSetupProperties.Options.supportedCardSchemes.rawValue] as? [String] {
            Fidel.supportedCardSchemes = CardScheme.cardSchemesSet(from: supportedCardSchemeJSKeys)
        }
        if options.keys.contains(FidelSetupProperties.Options.metaData.rawValue) {
            Fidel.metaData = options[FidelSetupProperties.Options.metaData.rawValue] as? [String: Any]
        }
    }

    private func setConsentTextValues(_ consentTextValues: [String: Any?]) {
        if consentTextValues.keys.contains(FidelSetupProperties.ConsentText.companyName.rawValue) {
            Fidel.companyName = consentTextValues[FidelSetupProperties.ConsentText.companyName.rawValue] as? String
        }
        if consentTextValues.keys.contains(FidelSetupProperties.ConsentText.termsAndConditionsURL.rawValue) {
            Fidel.termsAndConditionsURL = consentTextValues[FidelSetupProperties.ConsentText.termsAndConditionsURL.rawValue] as? String
        }
        if consentTextValues.keys.contains(FidelSetupProperties.ConsentText.privacyPolicyURL.rawValue) {
            Fidel.privacyPolicyURL = consentTextValues[FidelSetupProperties.ConsentText.privacyPolicyURL.rawValue] as? String
        }
        if consentTextValues.keys.contains(FidelSetupProperties.ConsentText.programName.rawValue),
           let programName = consentTextValues[FidelSetupProperties.ConsentText.programName.rawValue] as? String {
            Fidel.programName = programName
        }
        if consentTextValues.keys.contains(FidelSetupProperties.ConsentText.deleteInstructions.rawValue),
           let deleteInstructions = consentTextValues[FidelSetupProperties.ConsentText.deleteInstructions.rawValue] as? String {
            Fidel.deleteInstructions = deleteInstructions
        }
    }
}
